### DISCUSSION TITLE
Prometheus metrics with multiple profiles

### DISCUSSION THREAD

**Discussion by Javex** [creativeprojects/resticprofile#318] [Q&A] [answered]
Hi! Thank you for writing resticprofile, it's been an amazing tool in managing my restic backups <3

I recently enabled the `prometheus-save-to-file` option and started noticing some errors in my logs that look like this:

```
node_exporter[3734]: ts=2024-02-13T10:15:43.777Z caller=stdlib.go:105 level=error caller="error gathering metrics: [from Gatherer #2] collected metric \"resticprofile_build_info\" { label:{name:\"goversion\"  value:\"go1.21.3\"}  label:{name:\"version\"  value:\"0.24.0\"}  gauge:{value:1" msg="}} was collected before with the same name and label values"
```

After some investigation, I found that (I think) it has to do with having multiple profiles. In my simplest case I have two, as I have two repos. As such I have two separate `prometheus-save-to-file` options:

```
# ls /var/lib/node_exporter/textfile_collector/
restic-system-local.prom  restic-system.prom
```

And inside both files, most lines look like this:

```
# HELP resticprofile_backup_added_bytes Total number of bytes added to the repository.
# TYPE resticprofile_backup_added_bytes gauge
resticprofile_backup_added_bytes{profile="system"} 1.6002715e+08
```
That is, they have a `profile="system"` (and similarly a `profile="system-local"`) label.

But at the very end there is also this:

```
# HELP resticprofile_build_info resticprofile build information.
# TYPE resticprofile_build_info gauge
resticprofile_build_info{goversion="go1.21.3",version="0.24.0"} 1
```

This does not have a `profile` label and so when `node_exporter` looks at two `.prom` textfiles, it realises it's got two different files with the exact same metric here.

I'll be honest, I'm making this a Q&A/Discussion post rather than a bug report, because I have no idea who is "at fault" here... But I also can't think of how to work around this within `node_exporter` since it considers this an error-level issue and most of my search results have pointed to user error by the application developer (though not exactly cases like this, usually bugs).

The obvious solution is to add a `profile` label, but I don't know if that is the right way and I'm sure there was a reason not to include it.

As a workaround, I tried to simply use the _same_ file name for all profiles thinking that if `node_exporter` is quick enough it'll pick up the changes before the file gets overwritten. Unfortunately, this does not work out as some of my backups are rather quick and caused some gaps in metrics.

Are there any suggestions on what I might be doing wrong or is there something we could do to fix it?

Thanks again for all the hard work!

**Comment by creativeprojects** [comment, marked as answer]
Sorry I'm not using `node_exporter`...

Anyway I was trying with a local install, and playing with it a bit: I can confirm the `resticprofile_build_info` needs at least a different label so `node_exporter` can differentiate them.

You can test yourself by updating the text file and adding the same labels as the other metrics, and you'll see that `node-exporter` stops complaining 😉 

I'll fix this bug 👍🏻 

Thanks for reporting it 😛

**Reply by creativeprojects** [reply]
This PR https://github.com/creativeprojects/resticprofile/pull/319 should fix it.
You can give it a try from the branch if you feel confident doing so 😄

**Reply by Javex** [reply]
Thank you so much! And yeah haha I didn't think to just edit the file, that would have been an easy test >.<
